Adding doc for consolidation of multiple meshes (#938)
istio-ecosystem/sail-operator: docs/general/consolidating-cp.md

[Return to Project Root](../../README.md)

> # Consolidating control planes

Author: I'm not familiar with the documentation testing framework so I'm not sure about complexity. I can add it in another PR to unblock this one as it's needed for downstream documentation work.

> # Consolidating control planes

Reviewer: I can do it too, we can add the test in a separated PR without any issue

> It's possible to create multiple control planes in a single [cluster](https://istio.io/latest/docs/setup/install/multiple-controlplanes/). In case it's no longer needed to use multiple control planes, it might be useful to consolidate them to only one. This documentation shows how to achieve that without any traffic disruption.

Reviewer: ```suggestion
It's possible to create multiple control planes in a single [cluster](https://istio.io/latest/docs/setup/install/multiple-controlplanes/) using the Sail operator. For configuration details, refer to the documentation [here](https://github.com/istio-ecosystem/sail-operator/blob/main/docs/deployment-models/multiple-mesh.md#multiple-meshes-on-a-single-cluster). 
If you no longer need multiple control planes, it can be beneficial to consolidate them into a single one. This documentation outlines how to achieve this without any traffic disruption.
```

## Prerequisites

> By default, each control plane is using a different root certificate which prevents to move workloads from one mesh to another without breaking mTLS. In this procedure, we expect that each control plane is using an intermediate certificate issued by the same root CA. How to achieve this without any traffic disruption, see [Plug in CA Certificates](plugin-ca.md).

Reviewer: ```suggestion
By default, each control plane uses a different root certificate, which prevents moving workloads from one mesh to another without breaking mTLS. In this procedure, we expect that each control plane is using an intermediate certificate issued by the same root CA. To learn how to achieve this without any traffic disruption, see [Plug in CA Certificates](plugin-ca.md).
```

## Set up
We have two Istio control planes named `tenant-a` and `tenant-b` with following configuration:

> ```yaml

Reviewer: @FilipB, is there any configuration here that we need to remark? I mean, if there is any additional config that makes sense to remark will be good after showing the yaml, put some notes regarding that specific configuration

Author: No, only important part of the Istio CR is the `discoverySelectors` field

apiVersion: sailoperator.io/v1
kind: Istio
metadata:
...
  name: tenant-a
...
spec:
  namespace: istio-system-a
  updateStrategy:
    inactiveRevisionDeletionGracePeriodSeconds: 30
    type: RevisionBased
  values:
    meshConfig:
      defaultConfig:
        proxyMetadata:
          PROXY_CONFIG_XDS_AGENT: "true"
      discoverySelectors:
      - matchLabels:
          tenant: tenant-a
    pilot:
      env:
        ISTIO_MULTIROOT_MESH: "true"
  version: v1.24.5
status:
  activeRevisionName: tenant-a-v1-24-5
...
```
```yaml
apiVersion: sailoperator.io/v1
kind: Istio
metadata:
...
  name: tenant-b
...
spec:
  namespace: istio-system-b
  updateStrategy:
    inactiveRevisionDeletionGracePeriodSeconds: 30
    type: RevisionBased
  values:
    meshConfig:
      defaultConfig:
        proxyMetadata:
          PROXY_CONFIG_XDS_AGENT: "true"
      discoverySelectors:
      - matchLabels:
          tenant: tenant-a
    pilot:
      env:
        ISTIO_MULTIROOT_MESH: "true"
  version: v1.24.5
status:
  activeRevisionName: tenant-a-v1-24-5
...
```

> `sleep-a` and `httpbin-a` make up a `tenant-a` mesh and `sleep-b` and `httpbin-b` make up a `tenant-b` mesh. Namespaces are labeled with `istio.io/rev=tenant-a-v1-24-5` for injection and `tenant=tenant-a` or `tenant=tenant-b` for service discovery. Strict mTLS is enabled for both meshes.

Reviewer: ```suggestion
`sleep-a` and `httpbin-a` from the `tenant-a` mesh and `sleep-b` and `httpbin-b` from `tenant-b` mesh. Namespaces are labeled with `istio.io/rev=tenant-a-v1-24-5` for injection and `tenant=tenant-a` or `tenant=tenant-b` for service discovery. Strict mTLS is enabled for both meshes.
```

### Procedure
1. Make sure both control planes are using the same root certificate:
    ```bash
    kubectl get cm istio-ca-root-cert -n httpbin-a -o jsonpath={.data.'root-cert\.pem'}

>     kubectl get cm istio-ca-root-cert -n httpbin-b -o jsonpath={.data.'root-cert\.pem'}

Author: I don't  think there is a need for an example here. It will just print two same certificates
